python/lsst/pipe/base/pipeline.py

```diff
@@ -593,13 +590,43 @@ def _addConfigImpl(self, label: str, newConfig: pipelineIR.ConfigIR) -> None:
     def toFile(self, filename: str) -> None:
         self._pipelineIR.to_file(filename)
 
-    def write_to_uri(self, uri: Union[str, ButlerURI]) -> None:
-        # tasks need sorted each call because someone might have added or
-        # removed task, and caching changes does not seem worth the small
-        # overhead
-        labels = [td.label for td in self._toExpandedPipelineImpl(checkContracts=False)]
-        self._pipelineIR.reorder_tasks(labels)
-        self._pipelineIR.write_to_uri(uri)
+    def write_to_uri(
+        self,
+        uri: Union[str, ButlerURI],
+        expand: bool = False,
+        task_defs: Optional[Iterable[TaskDef]] = None,
+    ) -> None:
+        """Write the pipeline to a file or directory.
+
+        Parameters
+        ----------
+        uri : `str` or `ButlerURI`
+            URI to write to; may have any scheme with `ButlerURI` write
+            or no scheme for a local file/directory.  Should have a ``.yaml``
+            extension if ``expand=False`` and a trailing slash (indicating
+            a directory-like URI) if ``expand=True``.
+        expand : `bool`, optional
+            If `False`, write the pipeline to a single YAML file with
+            references to configuration files and other config overrides
+            unexpanded and unapplied, with tasks and subsets only minimally
+            validated (and not imported).  If `True`, import all tasks, apply
+            all configuration overrides (including those supplied by an
+            instrument), resolve parameters, sort all sections
+            deterministically, and write the pipeline to a directory with
+            separate config files for each task as well as a single
+            ``pipeline.yaml`` file.
+        task_defs : `Iterable` [ `TaskDef` ], optional
+            Output of `toExpandedPipeline`; may be passed to avoid a second
+            call to that method internally.  Ignored unless ``expand=True``.
+        """
+        if expand:
+            if str(uri).endswith(".yaml"):
+                raise RuntimeError(
+                    f"Expanded pipelines are written to directories, not YAML files like {uri}."
```

Comment:
I can work with `complete`; I don't think it's as evocative as `expanded`, but if that's in use already I agree that's a problem.  What do you think about "resolved"?  It's a bit less obvious what it means, but maybe that would help people from guessing what it means incorrectly.

On that note, we should probably still go looking for anywhere we call a pipeline with imports but not configs expanded an "expanded" pipeline and be sure to qualify it (and make sure we never say "fully expanded" in that context), because once this ticket lands (regardless of the terminology it uses), using "expanded" to mean something that isn't actually fully expanded is still going to be a source of confusion, because people _will_ guess incorrectly what that means.

Reply:
I'm fine with trying to migrate the language and or coming up with new terms, I just wanted to highlight what has been used and how people might get confused. I will think on what the best name will be, maybe even something like `finalized` or something like that.

Reply:
I like `finalized`; that also conveys the fact that I don't think people should be using this form if they want to apply more config overrides, even if we don't exactly prohibit it.

Reply:
So, I noticed a complication for changing `expand` -> `finalize`: `Pipeline` already has `toExpandedPipeline`, which _does_ mean essentially the same thing as what `expand` means on this branch (i.e. apply all config overrides).  Given that, I'm leaning more towards leaving this branch using `expand`, but looking for existing usage of `expand` where config override application is not implied, and adding qualifications to those.

Reply:
The only place I can think where it might be mentioned is in the pipeline documentation in pipe base. Other that that it will just be in the communications to people teaching them up to this point. I don't think it will be a huge deal, there might be some confusion from time to time. Can you think of a term for what we have now?